test/e2e/ftccontroller.go

```diff
@@ -0,0 +1,195 @@
+		}
+		existingCrd.Spec.Version = "v2"
+		existingCrd.Spec.Versions = []apiextv1b1.CustomResourceDefinitionVersion{
+			{
+				Name:    "v2",
+				Served:  true,
+				Storage: true,
+			},
+			{
+				Name:    "v1",
+				Served:  false,
+				Storage: false,
+			},
+		}
+		_, err = crdClient.CustomResourceDefinitions().Update(existingCrd)
+		if err != nil {
+			tl.Fatalf("Error updating target CRD version %q: %v", existingCrd.Spec.Version, err)
+		}
+
+		By("Federating the CRD v2")
+		needCleanup = false
+		typeConfig = enableResource(f, &targetAPIResource, "v2", needCleanup)
+
+		By("Waiting for refreshed sync controller ready")
+		objectMeta = typeConfig.GetObjectMeta()
+		waitForSyncerGeneration(tl, client, objectMeta.Namespace, objectMeta.Name)
+	})
+})
+
+func enableResource(f framework.KubeFedFramework, targetAPIResource *metav1.APIResource, version string, needCleanup bool) typeconfig.Interface {
+	tl := f.Logger()
+
+	enableTypeDirective := &kfenable.EnableTypeDirective{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: targetAPIResource.Name,
+		},
+		Spec: kfenable.EnableTypeDirectiveSpec{
+			TargetVersion:    version,
+			FederatedGroup:   targetAPIResource.Group,
+			FederatedVersion: targetAPIResource.Version,
+		},
+	}
+
+	resources, err := kfenable.GetResources(f.KubeConfig(), enableTypeDirective)
+	if err != nil {
+		tl.Fatalf("Error retrieving resources to enable federation of target type %q: %v", targetAPIResource.Kind, err)
+	}
+	typeConfig := resources.TypeConfig
+
+	err = kfenable.CreateResources(nil, f.KubeConfig(), resources, f.KubeFedSystemNamespace())
+	if err != nil {
+		tl.Fatalf("Error creating resources to enable federation of target type %q: %v", targetAPIResource.Kind, err)
+	}
+
+	if !needCleanup {
+		return typeConfig
+	}
+
+	framework.AddCleanupAction(func() {
+		delete := true
+		dryRun := false
+
+		objectMeta := typeConfig.GetObjectMeta()
+		qualifiedName := util.QualifiedName{Namespace: f.KubeFedSystemNamespace(), Name: objectMeta.Name}
+		err := kubefedctl.DisableFederation(nil, f.KubeConfig(), enableTypeDirective, qualifiedName, delete, dryRun, false)
+		if err != nil {
+			tl.Fatalf("Error disabling federation of target type %q: %v", targetAPIResource.Kind, err)
+		}
+	})
+	return typeConfig
+}
+
+func waitForTargetCrd(tl common.TestLogger, config *rest.Config, targetName, version string) {
+	err := wait.PollImmediate(framework.PollInterval, framework.TestContext.SingleCallTimeout, func() (bool, error) {
+		_, err := kfenable.LookupAPIResource(config, targetName, version)
+		if err != nil {
+			tl.Logf("An error was reported while waiting for target type %q to be published as an available resource: %v", targetName, err)
+		}
+		return (err == nil), nil
+	})
+	if err != nil {
+		tl.Fatalf("Timed out waiting for target type %q to be published as an available resource", targetName)
+	}
+}
+
+func waitForSyncerGeneration(tl common.TestLogger, client genericclient.Client, namespace, name string) {
```

Comment:
This name seemed a bit ambiguous to me at first. Consider renaming it to something that specifies waiting for the FTC generation to synchronize.